services/shhext/mailservers/connmanager_test.go

```diff
@@ -0,0 +1,301 @@
+	case <-time.After(time.Second):
+		require.FailNow(t, "can't send a drop event")
+	}
+	require.NoError(t, utils.Eventually(func() error {
+		nodes := server.Nodes()
+		if len(nodes) != target {
+			return fmt.Errorf("unexpected number of connected servers: %d", len(nodes))
+		}
+		if nodes[0] == initial {
+			return fmt.Errorf("connected node wasn't changed from %s", initial)
+		}
+		return nil
+	}, time.Second, 100*time.Millisecond))
+}
+
+func TestConnectionManagerReplace(t *testing.T) {
+	server := newFakeServer()
+	whisper := newFakeEnvelopesEvents()
+	target := 1
+	connmanager := NewConnectionManager(server, whisper, target)
+	connmanager.Start()
+	defer connmanager.Stop()
+	nodes := []*enode.Node{}
+	for _, n := range getNRandomNodes(t, 3) {
+		nodes = append(nodes, n)
+	}
+	connmanager.Notify(nodes[:1])
+	require.NoError(t, utils.Eventually(func() error {
+		connected := server.Nodes()
+		if len(connected) != target {
+			return fmt.Errorf("unexpected number of connected servers: %d", len(connected))
+		}
+		if nodes[0].ID() != connected[0] {
+			return fmt.Errorf("connected with a wrong peer. expected %s, got %s", nodes[0].ID(), connected[0])
+		}
+		return nil
+	}, time.Second, 100*time.Millisecond))
+	connmanager.Notify(nodes[1:])
+	require.NoError(t, utils.Eventually(func() error {
+		connected := server.Nodes()
+		if len(connected) != target {
+			return fmt.Errorf("unexpected number of connected servers: %d", len(connected))
+		}
+		switch connected[0] {
+		case nodes[1].ID():
+		case nodes[2].ID():
+		default:
+			return fmt.Errorf("connected with unexpected peer. got %s, expected %+v", connected[0], nodes[1:])
+		}
+		return nil
+	}, time.Second, 100*time.Millisecond))
+}
+
+func TestConnectionChangedAfterExpiry(t *testing.T) {
+	server := newFakeServer()
+	whisper := newFakeEnvelopesEvents()
+	target := 1
+	connmanager := NewConnectionManager(server, whisper, target)
+	connmanager.Start()
+	defer connmanager.Stop()
+	nodes := []*enode.Node{}
+	for _, n := range getNRandomNodes(t, 2) {
+		nodes = append(nodes, n)
+	}
+	connmanager.Notify(nodes)
+	var initial enode.ID
+	require.NoError(t, utils.Eventually(func() error {
+		nodes := server.Nodes()
+		if len(nodes) != target {
+			return fmt.Errorf("unexpected number of connected servers: %d", len(nodes))
+		}
+		initial = nodes[0]
+		return nil
+	}, time.Second, 100*time.Millisecond))
+	hash := common.Hash{1}
+	select {
+	case whisper.input <- whisperv6.EnvelopeEvent{
+		Event: whisperv6.EventMailServerRequestSent, Peer: initial, Hash: hash}:
+	case <-time.After(time.Second):
+		require.FailNow(t, "can't send a 'sent' event")
+	}
+	select {
+	case whisper.input <- whisperv6.EnvelopeEvent{
+		Event: whisperv6.EventMailServerRequestExpired, Peer: initial, Hash: hash}:
+	case <-time.After(time.Second):
+		require.FailNow(t, "can't send an 'expiry' event")
+	}
+	require.NoError(t, utils.Eventually(func() error {
```

Comment:
E.g. here a good comment would be 
```suggestion
	// Assert that peer got dropped as a result of envelop expiring
	require.NoError(t, utils.Eventually(func() error {
```